gc.c

```diff
@@ -362,7 +362,8 @@ sexp sexp_finalize (sexp ctx) {
       for (r=q->next; r && ((char*)r<(char*)p); q=r, r=r->next)
         ;
       if ((char*)r == (char*)p) { /* this is a free block, skip it */
-        p = (sexp) (((char*)p) + r->size);
+        // ISSUE_296 - Access to 'size' dereference NULL 'r'
+        p = (sexp) (((char*)p) + (r ? r->size : 0));
```

Comment:
r can easily be NULL if it's the last free block.  However, we're inside r == p, and p can never be NULL.  It's easy to see how the static analyzer might think this possible, though, but you must understand such tools are fallible.  If p ever were NULL here it would mean we have heap corruption, at which point all is lost and _nothing_ we do here matters.

You claim adding this check isn't "hurting" anything, but it is.  For one it's just that much extra code to read and understand.  Moreover, for the close reader it can add a huge amount of confusion - such a check suggests the heap may be partially corrupt yet we're expected to somehow recover from it here.  Except that the change in question doesn't recover, but just sets p to itself and enters an infinite loop.

Finally, this is in the inner loop of an inner loop that is scanning the entire heap - it is highly performance sensitive.  While there may be little difference on x86 with its branch prediction, it can be a significant slowdown on ARM.

This extra check helps nothing and hurts a lot.